app/src/behaviors/behavior_hold_tap.c

```diff
@@ -0,0 +1,491 @@
+	// [mt1_down, null, k1_up, mt2_up, null, ...]
+	//                  ^
+	// k1_up event is captured by the new hold-tap:
+	// [k1_down, k1_up, null, mt2_up, null, ...]
+	//                        ^
+	// mt2_up event is not captured but causes release of mt2 behavior
+	// [k1_down, k1_up, null, null, null, ...]
+	// now mt2 will start releasing it's own captured positions.
+	for (int i = 0; i < ZMK_BHV_HOLD_TAP_MAX_CAPTURED_EVENTS; i++) {
+		const struct zmk_event_header *captured_event = captured_events[i];
+		if (captured_event == NULL) {
+			return;
+		}
+		captured_events[i] = NULL;
+		if (undecided_hold_tap != NULL) {
+			k_msleep(10);
+		}
+		if (is_position_state_changed(captured_event)) {
+			struct position_state_changed *position_event = cast_position_state_changed(captured_event);
+			LOG_DBG("Releasing key position event for position %d %s", position_event->position, (position_event->state ? "pressed" : "released"));
+		} else {
+			struct keycode_state_changed *modifier_event = cast_keycode_state_changed(captured_event);
+			LOG_DBG("Releasing mods changed event 0x%02X %s", modifier_event->keycode, (modifier_event->state ? "pressed" : "released"));
+		}
+		ZMK_EVENT_RELEASE_AGAIN(captured_event);
+	}
+}
+
+
+/************************************************************    ACTIVE TAP HOLD HELPER FUNCTIONS */
+
+static struct active_hold_tap *find_hold_tap(u32_t position)
+{
+	for (int i = 0; i < ZMK_BHV_HOLD_TAP_MAX_HELD; i++) {
+		if (active_hold_taps[i].position == position) {
+			return &active_hold_taps[i];
+		}
+	}
+	return NULL;
+}
+
+static struct active_hold_tap *store_hold_tap(u32_t position, u32_t param_hold, u32_t param_tap, const struct behavior_hold_tap_config *config)
+{
+	for (int i = 0; i < ZMK_BHV_HOLD_TAP_MAX_HELD; i++) {
+		if (active_hold_taps[i].position != ZMK_BHV_HOLD_TAP_POSITION_NOT_USED) {
+			continue;
+		}
+		active_hold_taps[i].position = position;
+		active_hold_taps[i].is_decided = false;
+		active_hold_taps[i].is_hold = false;
+		active_hold_taps[i].config = config;
+		active_hold_taps[i].param_hold = param_hold;
+		active_hold_taps[i].param_tap = param_tap;
+		return &active_hold_taps[i];
+	}
+	return NULL;
+}
+
+static void clear_hold_tap(struct active_hold_tap *hold_tap)
+{
+	hold_tap->position = ZMK_BHV_HOLD_TAP_POSITION_NOT_USED;
+	hold_tap->is_decided = false;
+	hold_tap->is_hold = false;
+	hold_tap->work_is_cancelled = false;
+}
+
+enum decision_moment {
+	HT_KEY_UP               = 0,
+	HT_OTHER_KEY_DOWN       = 1,
+	HT_OTHER_KEY_UP         = 2,
+	HT_TIMER_EVENT          = 3,
+};
+
+static void decide_balanced(struct active_hold_tap *hold_tap, enum decision_moment event)
+{
+	switch (event) {
+	case HT_KEY_UP:
+		hold_tap->is_hold = 0;
+		hold_tap->is_decided = true;
+		break;
+	case HT_OTHER_KEY_UP:
+		hold_tap->is_hold = 1;
+		hold_tap->is_decided = true;
+		break;
+	case HT_TIMER_EVENT:
+		hold_tap->is_hold = 1;
+		hold_tap->is_decided = true;
+		break;
```

Comment:
Not sure on other opinions, but I like having stacked `case` statements for things like this:

```
case HT_OTHER_KEY_UP:
case HT_TIMER_EVENT:
   hold_tap->is_hold = 1;
   ...
```

Since its then obvious that both of those cases give an equivalent result, whereas having multiple cases makes me assume they will have differing effects